arelle/formula/FactAspectsCache.py

```diff
@@ -10,24 +10,60 @@
 
 
 class FactAspectsCache:
-    def __init__(self) -> None:
-        self._matchingAspects: defaultdict[ModelFact, defaultdict[ModelFact, defaultdict[int | QName, bool | None]]] = defaultdict(lambda: defaultdict(lambda: defaultdict(lambda: None)))
+    def __init__(self, maxSize: int) -> None:
+        self._size = 0
+        self._maxSize = maxSize if maxSize >= 0 else float("inf")
+        self._maxSizePreviouslyReached = False
+        self._prioritizedAspects: set[int | QName] = set()
+        self._matchingAspects: defaultdict[
+            ModelFact,
+            defaultdict[
+                ModelFact,
+                defaultdict[int | QName, bool | None]
+            ]
+        ] = defaultdict(lambda: defaultdict(lambda: defaultdict(lambda: None)))
 
-    def evaluations(self, fact1: ModelFact, fact2: ModelFact) -> defaultdict[int | QName, bool | None]:
-        return self._matchingAspects[fact1][fact2]
+    @property
+    def prioritizedAspects(self) -> set[int | QName]:
+        return self._prioritizedAspects
+
+    def evaluations(self, fact1: ModelFact, fact2: ModelFact) -> defaultdict[int | QName, bool | None] | None:
+        if not self._maxSizePreviouslyReached:
+            if fact1 in self._matchingAspects and fact2 in self._matchingAspects[fact1]:
+                return self._matchingAspects[fact1][fact2]
+            if fact2 in self._matchingAspects and fact1 in self._matchingAspects[fact2]:
+                return self._matchingAspects[fact2][fact1]
+        return None
 
     def cacheMatch(self, fact1: ModelFact, fact2: ModelFact, aspect: int | QName) -> None:
-        self._register(fact1, fact2, aspect, True)
+        if not self._maxSizePreviouslyReached:
+            self._register(fact1, fact2, aspect, True)
 
     def cacheNotMatch(self, fact1: ModelFact, fact2: ModelFact, aspect: int | QName) -> None:
-        self._register(fact1, fact2, aspect, False)
+        self._prioritizedAspects.add(aspect)
+        if not self._maxSizePreviouslyReached:
+            self._register(fact1, fact2, aspect, False)
 
     def _register(self, fact1: ModelFact, fact2: ModelFact, aspect: int | QName, value: bool) -> None:
+        if self._size >= self._maxSize:
+            # Disabling the cache entirely is somewhat rudimentary. Alternative caching strategies, such as LRU, were
+            # explored, but they demonstrated poorer performance on average across all tested documents. The current
```

Comment:
Just discussion, but this creates a surprising performance "cliff" where if a document gets to a specific size, it will suddenly get much worse performance.

Reply:
In practice, there isn't a sharp drop-off here. The truly massive, multi-dimensional documents end up with many unique values in the cache (that are never looked up), and the overhead of inserting values overtakes the benefit of the occasional cache hit. There's probably a better heuristic in detecting the number of dimensions, but I don't think it's worth the effort without a larger set of example documents to fine-tune the algorithm. I even tried disabling positive matches in the cache for large documents and continued inserting only false values (to enable short-circuiting). However, in the end, prioritized aspect sorting benefits these documents far more than the cached values. This is because, out of around 200 dimensions across 1000+ facts, all but 1-5 of them will match. If you can identify those 1-5 and check them first, the cache no longer provides a meaningful improvement.

I'll give Steven's suggestion a try and take another round of measurements, but I suspect the performance will be similar, with larger sustained memory usage.

On a related note, I initially implemented prioritized aspects using a counter to provide a more detailed weight for the sorting function. Surprisingly, through random experiments, I discovered that a simple set was more performant (because sorting was more expensive, or updating the counter, I don't know).

Reply:
I agree that for the docs+formulas you've looked at that there isn't a short drop-off.  I was thinking of a scenario with docs+formulas where the cache is effective and has 9M entries, and then the docs or formulas grow such that they would need a cache with 12M entries to hold them all.  Depending on what the formulas are doing, continuing to use the 10M cached entries could be significantly better than not using the cache at all, right?

Reply:
That's reasonably. It's entirely possible I'm over fitting the optimizations to the limited docs that I have. I went ahead and changed this to retain the cache.